README

	     =========================================================
 		      	    Geant4 - Hadrontherapy example
     	     =========================================================


                                README file
                           ----------------------


Last revision: May 2009;


                                    ACTUAL AUTHORS 

G.A.P. CIRRONE(*), G.CUTTONE, F. DI ROSA, F.ROMANO, G.RUSSO
M.RUSSO

Laboratori Nazionali del Sud - Istituto Nazionale di Fisica Nucleare
95123 Catania, Italy
* e-mail:cirrone@lns.infn.it


                                   PAST AUTHORS
S.GUATELLI, M.G. PIA
Istituto Nazionale di Fisica Nucleare, Sezione di Genova Via Dodecaneso, 33
16146, Genova, Italy

A. LECHNER (c)
CERN, Switzerland
16146, Genova, Italy


More informations on the Hadrontherapy example can be found in the 
Hadrontherapy Documentation available at
http://workgroup.lngs.infn.it/geant4lns/

Alternatevely send an e-mail to cirrone@lns.infn.it.


----> 0. INTRODUCTION.                                                    
                                                                       
The hadrontherapy example simulates an hadron-therapy beam line. 
In particular the example models the specific proton therapy beam line
installed at Laboratori Nazionali del Sud (INFN) in Catania, Sicily (Italy).
For more information on the proton therapy center of Catania
or/and proton/hadron therapy in general, please visit the
pages:
http://www.lns.infn.it/catanaweb/catana/ 

Moreover, since the Release of December 2008, with this example is also possible the 
simulation of carbon ion beams as the correct physic models are inserted with the
builted-in physic list denominated "QGSP_BIC".

---->1. GEOMETRY SET-UP.
 
The elements simulated are:

1. A scattering system, to spread geometrically the beam;

2. A system of collimators, to avoid the scattering radiation;

3. A modulation system that spreads the beam in energy and
   produces the so-called spread out bragg peak;
   It is constituted by a rotating wheel of different thichnesses.
   The wheel rotates around is axis (parallel to the proton
   beam axis) and its movement can be obtained by means of a 
   messenger between runs.

4. A set of monitor chambers (special transmission ionisation 
   chambers used to control hadron flux during the irradiation);

5. A "nozzle" and a final collimator defining the final shape
   of the beam before reaching the patient.

6. A water phantom: it is a box of water where the energy deposit is 
   calculated.
   The use of  the water phantom is required by the international protocol
   on the measure of dose in the case of proton and ion beams (IAEA 398, 2000).


-------> 2. PHYSICS

A particular care is addressed to the simulation of the physic processes.

Three differnt approaches can be used for the choose of the physic models.

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
APPROACH 1:
>>>>>>>>>>

Using the macro command 

/physic/addPhysics/<physics List name>.

In this case the models (for electromagnetic, hadronic elastic and hadronic inelstic) can be 
activated directly calling the name of the Physics Lists that are available inside the 
Geant4 kernel in the directory:

$G4INSTALL/source/physics_lists/builders/include 

An example of the use of the Physics List can be found in the macro files:
proton_therapy.mac and ion_therapy.mac

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
APPROACH 2:
>>>>>>>>>>

A set of built-in physic models are also contained inside the Hadrontherapy directory. These
are called Local*.cc and Local*.hh and can be activated using the macro command:
/physic/addPhysics/<name>.

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
APPROACH 3:
>>>>>>>>>>

We developed this approach in order to semplify the choice of the physic models to 
be used in the application.
**  With this approach the user must only insert a command line in his/her .mac file
    (using the /physics/addPackage PACKAGE_NAME).
    
This permits to switch-on an already build physic package.

** Various packages are already present in the Geant4 tree: they are in the directory:
   geant4/source/physics_lists/lists/include

** In this case hadronic inelastic models are directly activated for every particle.
*****************************************************************************************

------------------------------------------------------------------
SUGGESTED PHYSIC FOR ION BEAMS IN THE RANGE 0 - 400 MeV
------------------------------------------------------------------
Two macro files (proton_therapy.mac and ion_therapy.mac) can be used for proton and ion
simulations.

Also the QGSP_BIC package can be used if the *Approach 3* is preferred.


-------> INCL/ABLA physics lists

The application includes INCL/ABLA inelastic process setup for
deuteron, triton and alpha particles.

In order to start using generic ion support (up to Carbon) you must
first install an INCL/ABLA development version into your Geant4
setup. The code for this version can be found on GitHub
(github.com/kaitanie/incl).

------->2. EXPERIMENTAL SET-UP.      
                                      
The application simulates the proton therapy beam line
installed at Laboratori Nazionali del Sud.
The default beam line is a typical treatment line composed by several elements all
devoted to create the so-called "terapeutical beam", i.e. a beam ideal 
for a radiotherapeutic treatment.

   The main elements are:
** The COLLIMATORS: placed along the beam line to collimate the beam;

** The RANGE SHIFTERS: to decrease the energy of the primary proton beam 
   to a specific value;

** The MODULATOR WHEEL: to modulate the energy of the primary and monoenergetic
   beam in to a wide spectrum. The energy modulation is necessary to 
   homogeneusly irradiate a tumour volume that can extends in depth 
   up to 20 mm;

** The MONITOR CHAMBERS: very thin ionisation chamber that permit the 
   dose monitoring during the patient irradiation;

** The MOPI detector: microstrips, air free detector utilised for the check of the beam symmetry
   during the treatment

** The PATIENT COLLIMATOR: a brass, tumour-shaped collimator able to 
   confine the proton irradiation field in order to irradiate just the tumour mass
   in the trasverse direction;

The user has the possibility to vary, via messenger, almost all the geometrical 
characteristics of the beam line elements (i.e. their position along the beam line, 
their thickness, etc.). More details on the available user messengers can be 
found in the Hadronterapy Documentation: http://workgroup.lngs.infn.it/geant4lns/).

At the end of the beam line, a typical water phantom is reproduced.
A user-defined region of the phantom is divided (via the ROGeomtry class) in 
cubic and identical voxels. The voxels size can be varied. At the end of the simulation 
the energy deposited by primary protons, and secondaries in each voxel
is collected. This information is available as an .hbk file (if the 
G4ANALYSIS_USE variable is defined).  

The default sizes of the active voxelized region are 40x40x40 mm and actually the voxel configuration
is 200 x 1 x 1, which means 200 slices with 0.2 mm of thickness.Of course this default can be modified
in order to obtain, for example, a matrix of 80x80x80 cubic voxels each with a lateral dimension of 0.5 mm.

As concern the cut and stepMax values, the default configuration implies a cut value of 0.01 mm in the whole
 world (use the command /physic/setCuts 0.01 mm) and a stepMax of 0.01 mm just in the phantom 
(use the command /Step/waterPhantomStepMax 0.01 mm).
In any case it is strongly recommended to use a stepMax value not bigger 
than 5% of the dose slice thickness.

------->3. SET-UP 
                                                                        
- a standard Geant4 example GNUmakefile is provided                     

setup with:                                                             
compiler = gcc-3.4.6
G4SYSTEM = linux-g++                                                    

The following section reports the necessary environment variables 
necessary for the run of Hadrontherapy.                     

------->3.1  ENVIROMENT VARIABLES

 - G4SYSTEM = Linux-g++

 - G4INSTALL              points to the installation directory of GEANT4;

 - G4LIB                  point to the compiled libraries of GEANT4;

 - G4WORKDIR              points to the work directory;

 - CLHEP_BASE_DIR         points to the installation directory of CHLEP; 

 - G4LEVELGAMMADATA       points to the photoevaporation library;

 - G4NEUTRONHPDATA	  points to the neutron data files;
 
 - G4RADIOACTIVEDATA      points to the libraries for radio-active decay 
                          hadronic processes;

 - G4ABLADATA             points to the library of the INCL/ABLA hadronic model;
 
 - G4LEDATA               points to the low energy electromagnetic libraries

 - LD_LIBRARY_PATH = $CLHEP_BASE_DIR/lib:$LD_LIBRARY_PATH

---->3.2  VISUALISATION

The user can visualise the experimental set-up with OpenGL, DAWN and vrml

---->4. HOW TO RUN THE EXAMPLE                                         

In interactive mode:
> $G4WORDIR/bin/Linux-g++/Hadrontherapy
The defaultMacro.mac is executed

The primary particle beam parameter are:
Radiation:                proton beam;
Energy distribution:      gaussian;
Mean energy:              62.0 MeV;
Energy spread:            400 keV;


The modulator wheel can be rotated via the messenger:

Idle>/modulator/angle/xx deg

To produce a Spread Out Bragg Peak using the modulator a macro
(modulatorMacro.mac) is provided. With this macro the modulator is 
rotated of 360 degree at 1 deg steps. In each run 1000 protons are
generated as primary particles. Obviously a bigger resolution can be obtained 
with smaller angles or increasing the protons number in each run.

Modulator wheel can be omitted setting its material air.

run $G4WORKDIR/bin/Linux-g++/Hadrontherapy visualisationMacro.mac
to visualise the experimental set-up with OpenGL

---->6. SIMULATION OUTPUT                                    

The output is an .hbk file and a .root files. The files are  produced 
if the variable G4ANALYSIS_USE is set to 1 and the analysis tool (AIDA
interface) correctly installed.
The file contains  histograms and n-tuples.
The histograms contain the Bragg curves: energy deposited 
versus the depth in water (in mm) for the primary beam as well as all the secondaries produced. 
The n-tuple contains the total 3D energy deposit in the phantom; the information
is energy deposit in each voxel with respect to the position of the voxel.

Setup for analysis: AIDA 3.2.1  

Users can download the analysis tools from:  
http://aida.freehep.org/

Note that the same information can be stored in different format, 
like .root or .xml using the same AIDA interface.
Output is also stored in ASCII .out file. The file contains four columns; the first four columns represent the voxel indexs (that univocally identify the voxel volume) while the last column represent the energy deposited in the given voxel.


------->6.1  ROOT ANALYSIS

The application also supports using ROOT (http://root.cern.ch) directly without using AIDA.
This feature can be enabled by setting environment variable G4ROOTANALYSIS_USE to 1 and
unsetting G4ANALYSIS_USE.

Please contact cirrone@lns.infn.it for more details.

---------------------------------------------------------------------------

for comments, advices, doubts and questions please contact: 
cirrone@lns.infn.it



History
 -------------------------------------------------------------------------------
History File, 2004/02/27 G.A.P. Cirrone, Created
cirrone@lns.infn.it
http://g4advancedexamples.lngs.infn.it/Examples/hadrontherapy
-------------------------------------------------------------------------------

         ====================================================
             History file of the Hadrontherapy application
         ====================================================

20.11.2009 P.Kaitaniemi; Tag: hadrontherapy-V09-02-37
           - Fixes and updates to the analysis scripts in RootScripts/iaeaBenchmark
           - Updated RootScripts/README

18.11.2009 G.A.P.Cirrone; Tag: hadrontherapy-V09-02-36
           - Correction for a missing function in the HadrontherapyPhysicsListMessenger.cc class file

18.11.2009 G.A.P.Cirrone; Tag: hadrontherapy-V09-02-35
           - Updated the README file and general code revision for the 
             Geant4 9.3 release

17.11.2009 S.E.Mazzaglia; Tag: hadrontherapy-V09-02-34
	   - Added some functionalities in order to change, via messengers, the geometry, 
	     the voxelization of the detector, and the disposition in the space of the detector/phantom.
	   - Added the possibility to calculate the stopping powers for ions too.
	   - Modified the HadrontherapyDetectorROGeometry class constructor.
	   - Various additions and fixes to the matrix class.

10.11.2009 G.A.P.Cirrone; Tag: hadrontherapy-V09-02-33
           - Added the possibility to make a graphical user interface (GUI) using the QT libraries. 
             To start a GUI the correct enviroment variables must be configured (see the Geant4 installation 
             manual) and a QT version must be installed  

05.10.2009 P.Kaitaniemi; Tag: hadrontherapy-V09-02-32
           - Fixed a compilation error with GCC 4.4

28.09.2009 S.E.Mazzaglia; Tag: hadrontherapy-V09-02-31
           - Now the HadrontherapyDetectorConstruction class implements only phantom and detector [RO]geometry.
	     World volume and the rest of the geometry is inside another class whose messenger allows 
	     modification by users with the same old syntax
	   - Removed HadrontherapyInteractionParameters from the HadrontherapyGeometryController class 
			  
20.09.2009 P.Kaitaniemi; Tag: hadrontherapy-V09-02-30
           - Added ability to use command based scoring
           - IAEA geometry: produce Bragg peak using command based scoring
           - Various additions and fixes to the IAEA ROOT scripts
           - Additional data extracted from E. Haettner's thesis

20.09.2009 P.Kaitaniemi; Tag: hadrontherapy-V09-02-29
           - Moved HadrontherapyInteractionParameters initialization to the HadrontherapyGeometryController class

11.09.2009 G.A.P.Cirrone; Tag: hadrontherapy-V09-02-28
           - Added messengers to control the event number and to draw only particular tracks.
             The new command are accessible via the command /event/drawTracks and /event/PrintEventNumber  

08.09.2009 S.E.Mazzaglia; Tag: hadrontherapy-V09-02-27
           - Added a method to retrieve stopping power values for protons, alphas and electrons.
	     This method is implemented in the new class HadrontherapyInteractionParameters

13.08.2009 P.Kaitaniemi; Tag: hadrontherapy-V09-02-26
           - Fixed compilation errors when AIDA analysis is used

03.08.2009 P.Kaitaniemi; Tag: hadrontherapy-V09-02-25
           - Added ability to select the geometry using G4 macro commands
           - Improved plotting scripts and improved normalization for the
             fragment energy distribution

27.07.2009 P.Kaitaniemi; Tag: hadrontherapy-V09-02-24
           - IAEA geometry: added ability to remove the phantom by setting its thickness to zero
           - Collect simulation metadata: number of events, distance of the detector (IAEA geometry),
             depth of the phantom (IAEA geometry), beam energy, energy error
           - Added ability to produce angular distribution plots

17.07.2009 P.Kaitaniemi; Tag: hadrontherapy-V09-02-23
           - Tuned geometry of the E. Haettner experimet (IAEADetectorConstruction)
           - Adopted G4ANALYSIS_USE_ROOT flag to activate ROOT analysis
           - Improved plotting scripts

13.07.2009 P.Kaitaniemi; Tag: hadrontherapy-V09-02-22
           - Added the first version of the IAEA benchmark geometry based on
              E. Haettner's thesis
           - Collect fragment energy distributions
           - Added fragment energy distribution data
           - ROOT script preparing an IAEA benchmark figure with data

08.07.2009 G.A.P.Cirrone; Tag: hadrontherapy-V09-02-21
           - Removed the README file in ASCII format 

27.06.2009 G.A.P.Cirrone; Tag: hadrontherapy-V09-02-20
           - Eliminated not necessary dependences in the SteppingAction class
           - Added folders containing experimental data (its name is 'experimentalData') and 
             ROOT scripts ('RootScripts') where Root scripts are stored to
             perform a fast comparison with experimental data.
             A folder where simulation results are stored is also created. Its
             name is 'simulationResults'. 

27.06.2009 P. Kaitaniemi; Tag hadrontherapy-V09-02-19
	   - Added ability to change the name of the output file between runs

26.06.2009 P. Kaitaniemi; Tag hadrontherapy-V09-02-18
	   - Fixed a bug in the physics list. Local ion-ion hadronic physics was not loaded due
	     to an uninitialized variable (locIonIonInelasticIsRegistered)
	   - Ability to use /analysis/setAnalysisFile <filename> to set the name of the output file
	   - Added Doxygen documentation tags to the source code and Doxyfile for
	     documentation settings
	   - Support for direct use of ROOT for analysis in addition to the default AIDA one
	   - Local INCL/ABLA physics list for deuterons, tritons and alphas

26.06.2009 G.A.P.Cirrone; Tag hadrontherapy-V09-02-17

src/IAEADetectorConstruction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
// IAEADetectorConstruction.cc
//
// See more at: http://workgroup.lngs.infn.it/geant4lns/
//
// ----------------------------------------------------------------------------
//                 GEANT 4 - Hadrontherapy example
// ----------------------------------------------------------------------------
// Code developed by:
//
// G.A.P. Cirrone(a)*, F. Di Rosa(a), S. Guatelli(b), G. Russo(a)
// 
// (a) Laboratori Nazionali del Sud 
//     of the INFN, Catania, Italy
// (b) INFN Section of Genova
// 
// * cirrone@lns.infn.it
// ----------------------------------------------------------------------------
#include "G4SDManager.hh"
#include "G4RunManager.hh"
#include "G4Box.hh"
#include "G4LogicalVolume.hh"
#include "G4ThreeVector.hh"
#include "G4PVPlacement.hh"
#include "globals.hh"
#include "G4Transform3D.hh"
#include "G4RotationMatrix.hh"
#include "G4Colour.hh"
#include "G4UserLimits.hh"
#include "G4VisAttributes.hh"
#include "G4NistManager.hh"
#include "HadrontherapyAnalysisManager.hh"
#include "HadrontherapyDetectorROGeometry.hh"
#include "IAEADetectorMessenger.hh"
#include "HadrontherapyDetectorSD.hh"
#include "IAEADetectorConstruction.hh"
#include "HadrontherapyModulator.hh"

/////////////////////////////////////////////////////////////////////////////
IAEADetectorConstruction::IAEADetectorConstruction()
  : detectorSD(0), detectorROGeometry(0), 
    passiveProtonBeamLine(0), modulator(0),
    physicalTreatmentRoom(0),
    phantomPhysicalVolume(0), 
    detectorLogicalVolume(0), 
    detectorPhysicalVolume(0)
{
  // Messenger to change parameters of the geometry
  detectorMessenger = new IAEADetectorMessenger(this);

  // Detector sizes
  detectorSizeX = 20.*cm;
  detectorSizeY = 20.*cm;
  detectorSizeZ = 20.*cm;

  // Number of the detector voxels
  // This does notthing, since a primitive score is used  
  numberOfVoxelsAlongX = 400;
  numberOfVoxelsAlongY = 1;
  numberOfVoxelsAlongZ = 1;
  
  startDetectorThickness = 5.*cm; // approximation, exakt value not given by Haettner 2006
  phantomCenter = startDetectorThickness + 64.*cm;
  phantomDepth = 27.9 *cm;
  plexiThickness = 0.2 *cm;
  aluWindowThickness = 0.01 *cm;
  endDetectorThickness = 3.7 *cm;
  moveEndDetectorForward = 0 *cm; //if detector needs to be moved
  endDetectorPosition =  startDetectorThickness + 358 *cm + (endDetectorThickness / 2) - moveEndDetectorForward;
  
  noPhantom = false;
  
 }

/////////////////////////////////////////////////////////////////////////////
IAEADetectorConstruction::~IAEADetectorConstruction()
{ 
  if (detectorROGeometry) delete detectorROGeometry;  
  delete detectorMessenger;
}

/////////////////////////////////////////////////////////////////////////////
//void IAEADetectorConstruction::ChangeTheBeamLine(const G4String& name)
//if (name == emName) return;
//if (name == "ProtonPassiveBeamLine")
//{
//ConstructPassiveProtonBeamLine();
//}

/////////////////////////////////////////////////////////////////////////////
G4VPhysicalVolume* IAEADetectorConstruction::Construct()
{ 

  ConstructPassiveProtonBeamLine();
  ConstructDetector();
#ifdef ANALYSIS_USE
  //write the metadata for analysis
  HadrontherapyAnalysisManager::getInstance()->setGeometryMetaData(endDetectorPosition/10, phantomDepth/10, phantomCenter/10); //FIXME! unit correction hardcoded
#endif
  // Set the sensitive detector where the energy deposit is collected
  ConstructSensitiveDetector();
  return physicalTreatmentRoom;
}

/////////////////////////////////////////////////////////////////////////////
void IAEADetectorConstruction::ConstructPassiveProtonBeamLine()
{ 
  // -----------------------------
  // Treatment room - World volume
  //------------------------------
  // Treatment room sizes
  const G4double worldX = 400.0 *cm;
  const G4double worldY = 200.0 *cm; //to fit, new bigger detector
  const G4double worldZ = 200.0 *cm;
  G4bool isotopes = false;
 
  G4Material* airNist =  G4NistManager::Instance()->FindOrBuildMaterial("G4_AIR", isotopes);
  G4Box* treatmentRoom = new G4Box("TreatmentRoom",worldX,worldY,worldZ);
  G4LogicalVolume* logicTreatmentRoom = new G4LogicalVolume(treatmentRoom, 
                                                            airNist, 
                                                            "logicTreatmentRoom", 
															0,0,0);
  physicalTreatmentRoom = new G4PVPlacement(0,
					    G4ThreeVector(),
					    "physicalTreatmentRoom", 
					    logicTreatmentRoom, 
					    0,false,0);
 

  // The treatment room is invisible in the Visualisation
  logicTreatmentRoom -> SetVisAttributes (G4VisAttributes::Invisible);
 

 
  //----------------------------------------
  // Phantom:
  // A box used to approximate tissues. Is surrounded by plexi-glas.
  //----------------------------------------

  G4Material* waterNist = G4NistManager::Instance()->FindOrBuildMaterial("G4_WATER", isotopes);
  G4Material* plexiGlas = G4NistManager::Instance()->FindOrBuildMaterial("G4_PLEXIGLASS", isotopes);
  //G4Box* phantom = new G4Box("Phantom",10 *cm, 20 *cm, 20 *cm);
  //the below for integrated angular distribution plot
  G4Box* phantom = new G4Box("Phantom",phantomDepth/2, 32 *cm, 32 *cm); //< E.haettner has 32x8, but simulation takes scattering symmetrically
  G4Box* plexiSheet = new G4Box("phantomEdge",plexiThickness/2, 32 *cm, 32 *cm);
  G4LogicalVolume* phantomLogicalVolume = new G4LogicalVolume(phantom,	
							      waterNist, 
							      "phantomLog", 0, 0, 0);
								  
  G4LogicalVolume* phantomEdgeLogicalVolume = new G4LogicalVolume(plexiSheet,	
							      plexiGlas, 
							      "phantomEdgeLog", 0, 0, 0);
					//5.5cm for veto and start detector, see fig 5.1 and 4.1 ref Haettner 2006 

  if(!noPhantom){  
		  phantomPhysicalVolume = new G4PVPlacement(0,G4ThreeVector(phantomCenter, 0.*cm, 0.*cm),
								"phantomPhys",
								phantomLogicalVolume,
								physicalTreatmentRoom,
								false,0);

		  phantomEdge1PhysicalVolume = new G4PVPlacement(0,G4ThreeVector(phantomCenter - phantomDepth/2 - plexiThickness/2, 0.*cm, 0.*cm),
								"phantomEdgePhys",
								phantomEdgeLogicalVolume,
								physicalTreatmentRoom,
								false,0);
		  phantomEdge2PhysicalVolume = new G4PVPlacement(0,G4ThreeVector(phantomCenter + phantomDepth/2 + plexiThickness/2, 0.*cm, 0.*cm),
								"phantomEdgePhys",
								phantomEdgeLogicalVolume,
								physicalTreatmentRoom,
								false,0);
	}
  //----------------------------------------
  // Beamwindow:
  // The aluminium-window of the beam-source
  //----------------------------------------
  G4Material* aluNist = G4NistManager::Instance()->FindOrBuildMaterial("G4_Al", isotopes);
  G4Box* beamWindow = new G4Box("beamwindow",aluWindowThickness/2, 10 *cm, 10 *mm);
  G4LogicalVolume* beamWindowLogicalVolume = new G4LogicalVolume(beamWindow,	
							      aluNist, 
							      "beamWindowLog", 0, 0, 0);
  beamWindowPhysicalVolume = new G4PVPlacement(0,G4ThreeVector(1.*mm, 0.*mm, 0.*mm),
					    "beamPhys",
					    beamWindowLogicalVolume,
					    physicalTreatmentRoom,
					    false,0); //just ahead of beam that starts at origo
  //----------------------------------------
  // NewDetector:
  // A box used to simulate the end detector
  //----------------------------------------
  //G4Material* NewDetectorMaterial = G4NistManager::Instance()->FindOrBuildMaterial("G4_AIR", isotopes);
  //G4Box* NewDetector = new G4Box("NewDetector",1 *cm, 20 *cm, 27.9 *cm);
  
  // Visualisation attributes of the patient
  red = new G4VisAttributes(G4Colour(255/255., 0/255. ,0/255.));
  red -> SetVisibility(true);
  red -> SetForceSolid(true);
  //red -> SetForceWireframe(true);
  phantomLogicalVolume -> SetVisAttributes(red);
}

void IAEADetectorConstruction::setWaterThickness(G4double newWaterThickness){
	//This has to be run before the elements are made.
	if(newWaterThickness > 0){
	this->phantomDepth = newWaterThickness;
	}else{
	this->noPhantom = true;
	}	
	}


/////////////////////////////////////////////////////////////////////////////
void IAEADetectorConstruction::ConstructDetector()
{
  //-----------
  // Braggcurve Detector
  //-----------
  //Is currently not used
  
  G4bool isotopes =  false; 
  G4Material* waterNist = G4NistManager::Instance()->FindOrBuildMaterial("G4_WATER", isotopes);
  G4Box* detector = new G4Box("Detector",detectorSizeX,detectorSizeY,detectorSizeZ);

  detectorLogicalVolume = new G4LogicalVolume(detector,
					      waterNist,
					      "DetectorLog",
					      0,0,0);
  
  //G4double detectorXtranslation = -180.*mm;
  G4double detectorXtranslation = -180.*cm;
  detectorPhysicalVolume = new G4PVPlacement(0,
					     G4ThreeVector(detectorXtranslation, 0.0 *mm, 0.0 *mm),
					     "DetectorPhys",
					     detectorLogicalVolume,
					     phantomPhysicalVolume,
					     false,0);
  
  // Visualisation attributes of the phantom
  skyBlue = new G4VisAttributes( G4Colour(135/255. , 206/255. ,  235/255. ));
  skyBlue -> SetVisibility(true);
  skyBlue -> SetForceSolid(true);

  detectorLogicalVolume -> SetVisAttributes(skyBlue);
  
  //Visualization attributes for the beamwindow
  
  //-----------
  // NewDetector (mwpc etc. type behind the phantom)
  //-----------
  G4Material* NewDetectorMaterial = G4NistManager::Instance()->FindOrBuildMaterial("G4_AIR", false);
  G4Box* NewDetector = new G4Box("NewDetector",endDetectorThickness/2,190.*cm,190.*cm); //huge detector, will be scaled in root
  //For integrated angular distribution below
  NewDetectorLogicalVolume = new G4LogicalVolume(NewDetector,
					      NewDetectorMaterial,
					      "NewDetectorLog",
					      0,0,0);
  NewDetectorPhysicalVolume = new G4PVPlacement(0,
					     G4ThreeVector(endDetectorPosition, 0.0 *cm, 0.0 *cm),
					     "NewDetectorPhys",
					     NewDetectorLogicalVolume,
					     physicalTreatmentRoom,
					     false,0);
  
  
  // **************
  // Cut per Region
  // **************
  
  // A smaller cut is fixed in the phantom to calculate the energy deposit with the
  // required accuracy 
  // G4Region* aRegion = new G4Region("DetectorLog");
  //detectorLogicalVolume -> SetRegion(aRegion);
  //aRegion -> AddRootLogicalVolume(detectorLogicalVolume);
}
/////////////////////////////////////////////////////////////////////////////
void  IAEADetectorConstruction::ConstructSensitiveDetector()
{  
  // Sensitive Detector and ReadOut geometry definition
  G4SDManager* sensitiveDetectorManager = G4SDManager::GetSDMpointer();

  G4String sensitiveDetectorName = "Detector"; 
  if(!detectorSD)
    {
      // The sensitive detector is instantiated
      detectorSD = new HadrontherapyDetectorSD(sensitiveDetectorName);
      // The Read Out Geometry is instantiated
      G4String ROGeometryName = "DetectorROGeometry";
      detectorROGeometry = new HadrontherapyDetectorROGeometry(ROGeometryName,
							     phantomDepth/2,
							     detectorSizeY,
							     detectorSizeZ,
							     numberOfVoxelsAlongX,
							     numberOfVoxelsAlongY,
							     numberOfVoxelsAlongZ);
      detectorROGeometry -> BuildROGeometry();
      detectorSD -> SetROgeometry(detectorROGeometry);
      sensitiveDetectorManager -> AddNewDetector(detectorSD);
      detectorLogicalVolume -> SetSensitiveDetector(detectorSD);
    }
}
